File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (x, y) -> Integer.compare(x[0], y[0]));

        ArrayList<int[]> ans = new ArrayList<>();

        Arrays.stream(intervals).forEach(interval -> {
            if (ans.isEmpty() || interval[0] > ans.get(ans.size() - 1)[1]) {
                ans.add(interval);
            } else {
                ans.get(ans.size() - 1)[1] =
                    Math.max(ans.get(ans.size() - 1)[1], interval[1]);
            }
        });

        return ans.toArray(new int[ans.size()][]);
    }
}

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])

        ans = []
        prev_end = float('-inf')

        for start , end in intervals:
            if start > prev_end:
                ans.append([start, end])
                prev_end = end
            else:
                if prev_end < end:
                    ans[-1][1] = end
                    prev_end = end
        
        return ans